File: tests/algorithms/x/PHP/graphs/graph_adjacency_matrix.php
<?php
ini_set('memory_limit', '-1');
function _len($x) {
    if ($x === null) { return 0; }
    if (is_array($x)) { return count($x); }
    if (is_string($x)) { return strlen($x); }
    return strlen(strval($x));
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _panic($msg) {
    fwrite(STDERR, strval($msg));
    exit(1);
}
function make_graph($vertices, $edges, $directed) {
  $g = ['directed' => $directed, 'vertex_to_index' => [], 'adj_matrix' => []];
  $i = 0;
  while ($i < count($vertices)) {
  add_vertex($g, $vertices[$i]);
  $i = $i + 1;
};
  $j = 0;
  while ($j < count($edges)) {
  $e = $edges[$j];
  add_edge($g, $e[0], $e[1]);
  $j = $j + 1;
};
  return $g;
}
function contains_vertex($g, $v) {
  return isset($g['vertex_to_index'][$v]);
}
function add_vertex(&$g, $v) {
  if (contains_vertex($g, $v)) {
  _panic('vertex already exists');
}
  $matrix = $g['adj_matrix'];
  $i = 0;
  while ($i < count($matrix)) {
  $matrix[$i] = _append($matrix[$i], 0);
  $i = $i + 1;
};
  $row = [];
  $j = 0;
  while ($j < count($matrix) + 1) {
  $row = _append($row, 0);
  $j = $j + 1;
};
  $matrix = _append($matrix, $row);
  $g['adj_matrix'] = $matrix;
  $idx_map = $g['vertex_to_index'];
  $idx_map[$v] = count($matrix) - 1;
  $g['vertex_to_index'] = $idx_map;
}
function remove_key($m, $k) {
  global $g;
  $out = [];
  foreach (array_keys($m) as $key) {
  if ($key != $k) {
  $out[$key] = $m[$key];
}
};
  return $out;
}
function decrement_indices($m, $start) {
  global $g;
  $out = [];
  foreach (array_keys($m) as $key) {
  $idx = $m[$key];
  if ($idx > $start) {
  $out[$key] = $idx - 1;
} else {
  $out[$key] = $idx;
}
};
  return $out;
}
function remove_vertex(&$g, $v) {
  if (!contains_vertex($g, $v)) {
  _panic('vertex does not exist');
}
  $idx = ($g['vertex_to_index'])[$v];
  $new_matrix = [];
  $i = 0;
  while ($i < _len($g['adj_matrix'])) {
  if ($i != $idx) {
  $row = $g['adj_matrix'][$i];
  $new_row = [];
  $j = 0;
  while ($j < count($row)) {
  if ($j != $idx) {
  $new_row = _append($new_row, $row[$j]);
}
  $j = $j + 1;
};
  $new_matrix = _append($new_matrix, $new_row);
}
  $i = $i + 1;
};
  $g['adj_matrix'] = $new_matrix;
  $m = remove_key($g['vertex_to_index'], $v);
  $g['vertex_to_index'] = decrement_indices($m, $idx);
}
function add_edge(&$g, $u, $v) {
  if (!(contains_vertex($g, $u) && contains_vertex($g, $v))) {
  _panic('missing vertex');
}
  $i = ($g['vertex_to_index'])[$u];
  $j = ($g['vertex_to_index'])[$v];
  $matrix = $g['adj_matrix'];
  $matrix[$i][$j] = 1;
  if (!$g['directed']) {
  $matrix[$j][$i] = 1;
}
  $g['adj_matrix'] = $matrix;
}
function remove_edge(&$g, $u, $v) {
  if (!(contains_vertex($g, $u) && contains_vertex($g, $v))) {
  _panic('missing vertex');
}
  $i = ($g['vertex_to_index'])[$u];
  $j = ($g['vertex_to_index'])[$v];
  $matrix = $g['adj_matrix'];
  $matrix[$i][$j] = 0;
  if (!$g['directed']) {
  $matrix[$j][$i] = 0;
}
  $g['adj_matrix'] = $matrix;
}
function contains_edge($g, $u, $v) {
  if (!(contains_vertex($g, $u) && contains_vertex($g, $v))) {
  _panic('missing vertex');
}
  $i = ($g['vertex_to_index'])[$u];
  $j = ($g['vertex_to_index'])[$v];
  $matrix = $g['adj_matrix'];
  return $matrix[$i][$j] == 1;
}
function clear_graph(&$g) {
  $g['vertex_to_index'] = [];
  $g['adj_matrix'] = [];
}
$g = make_graph([1, 2, 3], [[1, 2], [2, 3]], false);
echo rtrim(_str($g['adj_matrix'])), PHP_EOL;
echo rtrim(_str(contains_edge($g, 1, 2))), PHP_EOL;
echo rtrim(_str(contains_edge($g, 2, 1))), PHP_EOL;
remove_edge($g, 1, 2);
echo rtrim(_str(contains_edge($g, 1, 2))), PHP_EOL;
remove_vertex($g, 2);
echo rtrim(_str($g['adj_matrix'])), PHP_EOL;

File: tests/algorithms/transpiler/FS/graphs/graph_adjacency_matrix.fs
// Generated 2025-08-15 11:16 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.15g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
type Graph = {
    mutable _directed: bool
    mutable _vertex_to_index: System.Collections.Generic.IDictionary<int, int>
    mutable _adj_matrix: int array array
}
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
open System.Collections.Generic

let rec make_graph (vertices: int array) (edges: int array array) (_directed: bool) =
    let mutable __ret : Graph = Unchecked.defaultof<Graph>
    let mutable vertices = vertices
    let mutable edges = edges
    let mutable _directed = _directed
    try
        let mutable g: Graph = { _directed = _directed; _vertex_to_index = _dictCreate<int, int> []; _adj_matrix = Array.empty<int array> }
        let mutable i: int = 0
        while i < (Seq.length (vertices)) do
            ignore (add_vertex (g) (_idx vertices (int i)))
            i <- i + 1
        let mutable j: int = 0
        while j < (Seq.length (edges)) do
            let e: int array = _idx edges (int j)
            ignore (add_edge (g) (_idx e (int 0)) (_idx e (int 1)))
            j <- j + 1
        __ret <- g
        raise Return
        __ret
    with
        | Return -> __ret
and contains_vertex (g: Graph) (v: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable g = g
    let mutable v = v
    try
        __ret <- (g._vertex_to_index).ContainsKey(v)
        raise Return
        __ret
    with
        | Return -> __ret
and add_vertex (g: Graph) (v: int) =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    let mutable g = g
    let mutable v = v
    try
        if contains_vertex (g) (v) then
            ignore (failwith ("vertex already exists"))
        let mutable matrix: int array array = g._adj_matrix
        let mutable i: int = 0
        while i < (Seq.length (matrix)) do
            matrix.[i] <- Array.append (_idx matrix (int i)) [|0|]
            i <- i + 1
        let mutable row: int array = Array.empty<int>
        let mutable j: int = 0
        while j < ((Seq.length (matrix)) + 1) do
            row <- Array.append row [|0|]
            j <- j + 1
        matrix <- Array.append matrix [|row|]
        g._adj_matrix <- matrix
        let mutable idx_map: System.Collections.Generic.IDictionary<int, int> = g._vertex_to_index
        idx_map <- _dictAdd (idx_map) (v) ((Seq.length (matrix)) - 1)
        g._vertex_to_index <- idx_map
        __ret
    with
        | Return -> __ret
and remove_key (m: System.Collections.Generic.IDictionary<int, int>) (k: int) =
    let mutable __ret : System.Collections.Generic.IDictionary<int, int> = Unchecked.defaultof<System.Collections.Generic.IDictionary<int, int>>
    let mutable m = m
    let mutable k = k
    try
        let mutable out: System.Collections.Generic.IDictionary<int, int> = _dictCreate []
        for key in m.Keys do
            if key <> k then
                out <- _dictAdd (out) (key) (_dictGet m (key))
        __ret <- out
        raise Return
        __ret
    with
        | Return -> __ret
and decrement_indices (m: System.Collections.Generic.IDictionary<int, int>) (start: int) =
    let mutable __ret : System.Collections.Generic.IDictionary<int, int> = Unchecked.defaultof<System.Collections.Generic.IDictionary<int, int>>
    let mutable m = m
    let mutable start = start
    try
        let mutable out: System.Collections.Generic.IDictionary<int, int> = _dictCreate []
        for key in m.Keys do
            let idx: int = _dictGet m (key)
            if idx > start then
                out <- _dictAdd (out) (key) (idx - 1)
            else
                out <- _dictAdd (out) (key) (idx)
        __ret <- out
        raise Return
        __ret
    with
        | Return -> __ret
and remove_vertex (g: Graph) (v: int) =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    let mutable g = g
    let mutable v = v
    try
        if not (contains_vertex (g) (v)) then
            ignore (failwith ("vertex does not exist"))
        let idx: int = _dictGet (g._vertex_to_index) (v)
        let mutable new_matrix: int array array = Array.empty<int array>
        let mutable i: int = 0
        while i < (Seq.length (g._adj_matrix)) do
            if i <> idx then
                let mutable row: int array = _idx (g._adj_matrix) (int i)
                let mutable new_row: int array = Array.empty<int>
                let mutable j: int = 0
                while j < (Seq.length (row)) do
                    if j <> idx then
                        new_row <- Array.append new_row [|(_idx row (int j))|]
                    j <- j + 1
                new_matrix <- Array.append new_matrix [|new_row|]
            i <- i + 1
        g._adj_matrix <- new_matrix
        let mutable m: System.Collections.Generic.IDictionary<int, int> = remove_key (g._vertex_to_index) (v)
        g._vertex_to_index <- decrement_indices (m) (idx)
        __ret
    with
        | Return -> __ret
and add_edge (g: Graph) (u: int) (v: int) =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    let mutable g = g
    let mutable u = u
    let mutable v = v
    try
        if not ((contains_vertex (g) (u)) && (contains_vertex (g) (v))) then
            ignore (failwith ("missing vertex"))
        let mutable i: int = _dictGet (g._vertex_to_index) (u)
        let mutable j: int = _dictGet (g._vertex_to_index) (v)
        let mutable matrix: int array array = g._adj_matrix
        matrix.[i].[j] <- 1
        if not (g._directed) then
            matrix.[j].[i] <- 1
        g._adj_matrix <- matrix
        __ret
    with
        | Return -> __ret
and remove_edge (g: Graph) (u: int) (v: int) =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    let mutable g = g
    let mutable u = u
    let mutable v = v
    try
        if not ((contains_vertex (g) (u)) && (contains_vertex (g) (v))) then
            ignore (failwith ("missing vertex"))
        let mutable i: int = _dictGet (g._vertex_to_index) (u)
        let mutable j: int = _dictGet (g._vertex_to_index) (v)
        let mutable matrix: int array array = g._adj_matrix
        matrix.[i].[j] <- 0
        if not (g._directed) then
            matrix.[j].[i] <- 0
        g._adj_matrix <- matrix
        __ret
    with
        | Return -> __ret
and contains_edge (g: Graph) (u: int) (v: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable g = g
    let mutable u = u
    let mutable v = v
    try
        if not ((contains_vertex (g) (u)) && (contains_vertex (g) (v))) then
            ignore (failwith ("missing vertex"))
        let mutable i: int = _dictGet (g._vertex_to_index) (u)
        let mutable j: int = _dictGet (g._vertex_to_index) (v)
        let mutable matrix: int array array = g._adj_matrix
        __ret <- (_idx (_idx matrix (int i)) (int j)) = 1
        raise Return
        __ret
    with
        | Return -> __ret
and clear_graph (g: Graph) =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    let mutable g = g
    try
        g._vertex_to_index <- _dictCreate<int, int> []
        g._adj_matrix <- Array.empty<int array>
        __ret
    with
        | Return -> __ret
let mutable g: Graph = make_graph (unbox<int array> [|1; 2; 3|]) ([|[|1; 2|]; [|2; 3|]|]) (false)
ignore (printfn "%s" (_str (g._adj_matrix)))
ignore (printfn "%s" (_str (contains_edge (g) (1) (2))))
ignore (printfn "%s" (_str (contains_edge (g) (2) (1))))
ignore (remove_edge (g) (1) (2))
ignore (printfn "%s" (_str (contains_edge (g) (1) (2))))
ignore (remove_vertex (g) (2))
ignore (printfn "%s" (_str (g._adj_matrix)))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
